posts/2018-06-15-typedapi.md

```diff
@@ -0,0 +1,329 @@
+  * `method` should be obvious. It takes some method type.
+  * `PathList` is a type carrier with a function `def /(...)` to concatenate path elements and segments. In the end, `PathList` only stores the type of an `HList` and nothing more.
+  
+```scala
+final case class PathList[P <: HList]() {
+  
+  def /[S](path: Witness.Lt[S]): PathList[S :: P] = PathList()
+  ...
+}
+
+val Root = PathList[HNil]()
+```
+  * Same is true for `QueryList`.
+  * The last step is to merge all these `HLists` types into a single one. Shapeless comes again with a handy type class called `Prepend` which provides us with the necessary functionality. Two `HList` types go in, a single type comes out. And again, we use a type carrier here to store the api type.
+
+Whoho, we did it. One thing we can mark as done on our todo list. Next step is to derive an actual client function from it.
+
+### Clients from types
+So far we have a type carrier with our api:
+
+```scala
+ApiTypeCarrier[Get[List[User]] :: Query[minAgeW.T, Int] :: Segment[nameW.T, String] :: usersW.T :: HNil]
+```
+
+Now we want to transform that into a function call `(name: String, minAge: Int) => F[List[User]]`. So what we need is the following:
+  * the types of our expected input
+  * the output type
+  * the path to the endpoint we want to call
+
+All information are available but mixed up and we need to separate them. Usually, when we work with collections and want to change their shape we do a `fold` and alas shapeless has type classes to fold left and right over an `HList`. But we only have a type. How do we fold that?
+
+#### Type-level FoldLeft
+What we want is to go from `Api <: HList` to `(El <: HList, KIn <: HList, VIn <: HList, M, Out)` with:
+  * `El` al the elements in our api: `"users".type :: SegmentInput :: QueryInput :: GetCall :: HNil`
+  * `KIn` the input key types: `nameW.T :: minAgeW.T :: HNil`
+  * `VIn` the input value types: `String :: Int :: HNil`
+  * the method type: `GetCall`
+  * and `Out`: `List[User]`
+
+Here, we introduced new types `SegmentInput` and `QueryInput` which act as placeholders and indicate that our api has the following inputs. This representation will come in handy when we construct our function.
+
+Now, how to fold on the type-level? The first step, we have to define a function which describes how to aggregate two types:
+
+```scala
+trait FoldLeftFunction[In, Agg] { type Out }
+```
+
+That's it. We say what goes in and what comes out. You need some examples to get a better idea? Here you go:
+
+```scala
+implicit def pathTransformer[P, El <: HList, KIn <: HList, VIn <: HList, M, Out] = 
+  FoldLeftFunction[Path[P], (El, KIn, VIn, M, Out)] { type Out = (P :: El, KIn, VIn, Out) }
+
+implicit def segmentTransformer[K <: Symbol, V, El <: HList, KIn <: HList, VIn <: HList, M, Out] = 
+  FoldLeftFunction[Segment[K, V], (El, KIn, VIn, M, Out)] { type Out = (SegmentInput :: El, K :: KIn, V :: VIn, Out) }
+
```

Comment:
A short explanation in English might help here, for people who're not used to the Scala typelevel programming idioms